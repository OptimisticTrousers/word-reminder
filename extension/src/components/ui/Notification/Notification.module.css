.notification {
  border-radius: 6px;
  padding: 1rem;
}

.notification--success {
  --bkg-color: rgb(220, 252, 231);
  --base-text-color: rgb(20, 83, 45);
  --btn-text-color: rgb(22, 163, 74);
  --icon-color: rgb(21, 128, 61);
  background-color: var(--bkg-color);
}

.notification--error {
}

.notification__flex {
  display: flex;
  align-items: center;
}

.notification__item {
  flex-shrink: 0;
}

.notification__box {
  margin-left: 0.75rem;
}

.notification__message {
  color: var(--base-text-color);
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.25rem;
  display: block;
}

.notification__controls {
  margin-left: auto;
  padding-left: 0.75rem;
}

.notification__control {
  margin-top: -0.375rem;
  margin-right: -0.375rem;
  margin-bottom: -0.375rem;
  margin-left: -0.375rem;
}

.notification__button {
  color: var(--btn-text-color);
  padding: 0.375rem;
  background-color: var(--bkg-color);
  border-radius: 6px;
  display: inline-flex;
}

.notification__text {
  position: absolute;
  width: 1px;
  padding: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border-width: 0;
}

.notification__icon {
  width: 1.25rem;
  height: 1.25rem;
}

.notification__icon--circle {
  color: var(--icon-color);
}
