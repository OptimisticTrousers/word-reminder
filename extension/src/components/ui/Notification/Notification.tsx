import { CircleCheck, CircleX, X } from "lucide-react";
import { useContext } from "react";
import CSSModules from "react-css-modules";

import {
  NOTIFICATION_ACTIONS,
  NotificationContext,
} from "../../../context/Notification";
import styles from "./Notification.module.css";

export const Notification = CSSModules(
  function ({ type, message }: { type: string; message: string }) {
    const { dismissNotification } = useContext(NotificationContext);

    function handleClick() {
      dismissNotification();
    }

    return (
      <div
        styleName={`notification notification--${type}`}
        role="dialog"
        aria-labelledby="notification-message"
      >
        <div styleName="notification__flex">
          <div styleName="notification__item">
            {type === NOTIFICATION_ACTIONS.SUCCESS ? (
              <CircleCheck
                data-testid="check-icon"
                styleName="notification__icon notification__icon--circle"
              />
            ) : (
              <CircleX
                data-testid="x-icon"
                styleName="notification__icon notification__icon--circle"
              />
            )}
          </div>
          <div styleName="notification__box">
            <span id="notification-message" styleName="notification__message">
              {message}
            </span>
          </div>
          <div styleName="notification__controls">
            <div styleName="notification__control">
              <button styleName="notification_button" onClick={handleClick}>
                <span styleName="notification__text">Dismiss</span>
                <X styleName="notification__icon" />
              </button>
            </div>
          </div>
        </div>
      </div>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);
