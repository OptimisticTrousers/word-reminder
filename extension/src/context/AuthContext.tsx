import { createContext, FC, useReducer } from "react";

interface IUser {
  _id: string;
  username: string;
  theme: string;
  words: IUserWord[];
}

interface IWord {
  _id: string;
  word: string;
  origin: string;
  phonetic: string;
  audio: string;
  meanings: [
    {
      partOfSpeech: string;
      definitions: [
        {
          definition: string;
          example: string;
          synonyms: string[];
          antonyms: string[];
        }
      ];
    }
  ];
}

export interface IUserWord {
  _id: string;
  userId: IUser | string;
  word: IWord;
  learned: boolean;
}

interface AuthState {
  user: IUser | null;
}

interface Action {
  type: string;
  payload: IUser;
}

interface AuthContext {
  state: AuthState;
  dispatch: (action: Action) => void;
}

const authReducer = (state: AuthState, action: Action) => {
  switch (action.type) {
    // Login action will always contain a payload of user object
    case "LOGIN":
      return { user: action.payload };
    // Remove user object on logout action dispatch
    case "LOGOUT":
      return { user: null };
    default:
      return state;
  }
};

export const AuthContext = createContext<AuthContext>({} as AuthContext);

interface Props {
  children: JSX.Element[] | JSX.Element;
}

const AuthProvider: FC<Props> = ({ children }) => {
  const [state, dispatch] = useReducer(authReducer, {
    user: null,
  });

  return (
    <AuthContext.Provider value={{ state, dispatch }}>
      {children}
    </AuthContext.Provider>
  );
};

export default AuthProvider;
