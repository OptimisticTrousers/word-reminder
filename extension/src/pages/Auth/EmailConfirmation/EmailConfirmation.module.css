.email-confirmation {
  padding: 1rem;
}

.email-confirmation__heading {
  margin-bottom: 0.75rem;
}

.email-confirmation__bold {
  font-weight: 600;
}
