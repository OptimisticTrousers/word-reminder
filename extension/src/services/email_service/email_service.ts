import { Templates } from "common";

import { service } from "../service";

interface Body {
  email: string;
  subject: string;
  template: Templates;
}

export const emailService = (function (service) {
  const { post, VITE_API_DOMAIN } = service;

  function sendEmail(body: Body) {
    return post({
      url: `${VITE_API_DOMAIN}/emails`,
      options: {
        body: JSON.stringify(body),
        credentials: "include",
      },
    });
  }

  return { sendEmail };
})(service);
