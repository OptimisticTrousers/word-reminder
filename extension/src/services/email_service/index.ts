export { emailService } from "./email_service";
