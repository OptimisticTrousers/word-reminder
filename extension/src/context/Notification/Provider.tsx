import { useReducer, ReactNode } from "react";

import { NOTIFICATION_ACTIONS, NotificationContext } from "./Context";
import { Notification } from "../../components/ui/Notification/Notification";

interface Props {
  children: ReactNode;
}

interface NotificationState {
  isVisible: boolean;
  type: string;
  message: string;
}

interface NotificationAction {
  payload: { message: string };
  type: string;
}

export interface NotificationContext {
  showNotification: (type: string, message: string) => void;
  dismissNotification: () => void;
}

const notificationReducer = (
  state: NotificationState,
  action: NotificationAction
) => {
  switch (action.type) {
    case NOTIFICATION_ACTIONS.SUCCESS:
      return {
        ...state,
        isVisible: true,
        type: NOTIFICATION_ACTIONS.SUCCESS,
        message: action.payload.message,
      };
    case NOTIFICATION_ACTIONS.ERROR:
      return {
        ...state,
        isVisible: true,
        type: NOTIFICATION_ACTIONS.ERROR,
        message: action.payload.message,
      };
    case DISAPPEAR:
      return { ...state, isVisible: false, type: "", message: "" };
    default: {
      throw Error("Unknown action: " + action.type);
    }
  }
};

// Provide ability to spawn Toast notification globally within the application
export function NotificationProvider({ children }: Props) {
  const [state, dispatch] = useReducer(notificationReducer, {
    isVisible: false,
    type: "",
    message: "",
  });

  function dismissNotification() {
    dispatch({
      type: DISAPPEAR,
      payload: { message: "" },
    });
  }

  function showNotification(type: string, message: string): void {
    const delay = 5000;

    dispatch({ type, payload: { message } });

    setTimeout(() => {
      dismissNotification();
    }, delay);
  }

  const { isVisible, type, message } = state;

  return (
    <NotificationContext.Provider
      value={{ showNotification, dismissNotification }}
    >
      {isVisible && <Notification type={type} message={message} />}
      {children}
    </NotificationContext.Provider>
  );
}

const DISAPPEAR = "disappear";
