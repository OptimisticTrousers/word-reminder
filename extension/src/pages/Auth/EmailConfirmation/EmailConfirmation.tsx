import { Subject, Templates } from "common";
import CSSModules from "react-css-modules";
import { useQuery } from "@tanstack/react-query";

import { emailService } from "../../../services/email_service";
import styles from "./EmailConfirmation.module.css";
import { Loading } from "../../../components/ui/Loading";
import { Error500 } from "../../Error500";

interface Props {
  email: string;
}

export const EmailConfirmation = CSSModules(
  function ({ email }: Props) {
    const { isLoading, failureReason } = useQuery({
      queryKey: ["emails"],
      queryFn: () => {
        return emailService.sendEmail({
          email: email,
          subject: Subject.CHANGE_VERIFICATION,
          template: Templates.CONFIRM_EMAIL,
        });
      },
    });

    if (failureReason) {
      return <Error500 message={failureReason.message} />;
    }

    if (isLoading) {
      return <Loading />;
    }

    return (
      <div styleName="email-confirmation">
        <h2 styleName="email-confirmation__heading">Check your email</h2>
        <p styleName="email-confirmation__message">
          Follow the link in the email sent to
          <span styleName="email-confirmation__bold"> {email} </span>
          and continue creating your account.
        </p>
      </div>
    );
  },
  styles,
  { allowMultiple: true, handleNotFoundStyleName: "log" }
);
