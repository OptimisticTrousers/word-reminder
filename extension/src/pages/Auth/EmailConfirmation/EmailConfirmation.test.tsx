import { render, screen } from "@testing-library/react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

import { emailService } from "../../../services/email_service";
import { EmailConfirmation } from "./EmailConfirmation";
import { Subject, Templates } from "common";
import { NotificationProvider } from "../../../context/Notification";
import { MemoryRouter } from "react-router-dom";

vi.mock("../../../components/ui/Loading/Loading");

vi.mock("../../Error500/Error500");

describe("EmailConfirmation component", () => {
  const email = "bob@gmail.com";

  const info = {
    message:
      "Message sent: <d786aa62-4e0a-070a-47ed-0b0666549519@ethereal.email",
  };

  const status = 200;

  function setup() {
    const queryClient = new QueryClient();

    return render(
      <NotificationProvider>
        <QueryClientProvider client={queryClient}>
          <MemoryRouter>
            <EmailConfirmation email={email} />
          </MemoryRouter>
        </QueryClientProvider>
      </NotificationProvider>
    );
  }

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("calls the functions to send confirmation email", async () => {
    const mockSendEmail = vi
      .spyOn(emailService, "sendEmail")
      .mockImplementation(async () => {
        return { json: { info }, status };
      });

    const { asFragment } = setup();

    const heading = await screen.findByRole("heading", {
      name: "Check your email",
    });
    const message = screen.getByText((_, element) => {
      return (
        element?.textContent ===
        `Follow the link in the email sent to ${email} and continue creating your account.`
      );
    });
    expect(heading).toBeInTheDocument();
    expect(message).toBeInTheDocument();
    expect(mockSendEmail).toHaveBeenCalledTimes(1);
    expect(mockSendEmail).toHaveBeenCalledWith({
      email: email,
      subject: Subject.CHANGE_VERIFICATION,
      template: Templates.CONFIRM_EMAIL,
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it("shows that it is sending email when loading", async () => {
    const delay = 50;
    const mockSendEmail = vi
      .spyOn(emailService, "sendEmail")
      .mockImplementation(async () => {
        return new Promise((resolve) => {
          setTimeout(() => {
            resolve({ json: { info }, status });
          }, delay);
        });
      });

    setup();

    const loading = screen.getByTestId("loading");
    expect(loading).toBeInTheDocument();
    expect(mockSendEmail).toHaveBeenCalledTimes(1);
    expect(mockSendEmail).toHaveBeenCalledWith({
      email: email,
      subject: Subject.CHANGE_VERIFICATION,
      template: Templates.CONFIRM_EMAIL,
    });
  });

  it("shows that it is sending email when there is an error", async () => {
    const message = "Bad Request.";
    const status = 400;
    const mockSendEmail = vi
      .spyOn(emailService, "sendEmail")
      .mockImplementation(async () => {
        return Promise.reject({ json: { message }, status });
      });

    setup();

    const error = await screen.findByTestId("error-500");
    expect(error).toBeInTheDocument();
    expect(mockSendEmail).toHaveBeenCalledTimes(1);
    expect(mockSendEmail).toHaveBeenCalledWith({
      email: email,
      subject: Subject.CHANGE_VERIFICATION,
      template: Templates.CONFIRM_EMAIL,
    });
  });
});
