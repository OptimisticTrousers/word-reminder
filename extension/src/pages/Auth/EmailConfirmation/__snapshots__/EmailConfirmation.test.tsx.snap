// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EmailConfirmation component > calls the functions to send confirmation email 1`] = `
<DocumentFragment>
  <div
    class="_email-confirmation_9d1292"
  >
    <h2
      class="_email-confirmation__heading_9d1292"
    >
      Check your email
    </h2>
    <p
      class="_email-confirmation__message_9d1292"
    >
      Follow the link in the email sent to
      <span
        class="_email-confirmation__bold_9d1292"
      >
         bob@gmail.com 
      </span>
      and continue creating your account.
    </p>
  </div>
</DocumentFragment>
`;
